Add template for SingleR evaluation (#634)
AlexsLemonade/OpenScPCA-analysis: analyses/cell-type-ewings/template_notebooks/singler-workflow/01-singler-results.Rmd

  - [`SILIGAN_TARGETS_OF_EWS_FLI1_FUSION_DN`](https://www.gsea-msigdb.org/gsea/msigdb/cards/SILIGAN_TARGETS_OF_EWS_FLI1_FUSION_DN)

For both marker gene expression and gene set scores, the following plots will be shown: 

- Heatmap showing gene expression or gene set scores across all cells. 
Here the rows correspond to the gene list and the columns are the cells. 
The annotations are shown below the heatmap. 
- Density plot showing gene expression or gene set scores across all cells. 
Each row is a cell type and the expression or score is plotted on the x-axis. 

### Tumor vs. Normal 

In this section we show just the cells that are considered tumor and normal, lumping all non-tumor cell types together. 

**Marker gene expression**

```{r}
full_celltype_heatmap(classification_df, marker_gene_columns, "singler_tumor_normal")
```

```{r}
plot_density(classification_df,
             "tumor_sum",
             "singler_tumor_normal")
```

**Gene set scores**

```{r}
full_celltype_heatmap(classification_df, geneset_columns, "singler_tumor_normal")
```

```{r}

> geneset_columns |> 
>   purrr::map(\(column){

Reviewer: Just noting that the code here is writing out the list names, which I think can be avoided by capturing the `map` output:
```suggestion
plots <- geneset_columns |> 
  purrr::map(\(column){
```

I don't _think_ that would prevent the plots from printing, but if it did you could use `patchwork:: wrap_plots()` on the resulting list.

  "Marker genes file does not exist" = file.exists(params$marker_genes_file)
)
```

```{r packages}
suppressPackageStartupMessages({
  # load required packages
  library(SingleCellExperiment)
  library(ggplot2)
})

# Set default ggplot theme
theme_set(
  theme_bw()
)

# quiet messages
options(readr.show_col_types = FALSE)
ComplexHeatmap::ht_opt(message = FALSE)
```

```{r base paths}
# The path to this module
module_base <- rprojroot::find_root(rprojroot::is_renv_project)

# source in helper functions: plot_gene_heatmap() and plot_cnv_heatmap() 
# create_classification_df() and create_marker_gene_df()
validation_functions <- file.path(module_base, "scripts", "utils", "tumor-validation-helpers.R")
source(validation_functions)
```

```{r}
# Read in files
sce <- readr::read_rds(params$sce_file)
singler_results_df <- readr::read_tsv(params$singler_results_file)
geneset_scores_df <- readr::read_tsv(params$geneset_scores_file)
```

## Prep for plots

```{r}
# original library annotation 
orig_tumor <- glue::glue("tumor-{params$library_id}")

# generate classification df to use for plots 
classification_df <- sce |> 
  scuttle::makePerCellDF(use.dimred = "UMAP") |> 

>   # replace UMAP.1 with UMAP1
>   dplyr::rename_with(
>     \(x) stringr::str_replace(x, "^UMAP\\.", "UMAP")
>   ) |>
>   # get rid of excess columns
>   dplyr::select(barcodes, UMAP1, UMAP2) |> 

Reviewer: I don't know why I haven't suggested this before, but maybe could to simplify this to:
```suggestion
  # get rid of excess columns and rename UMAPs
  dplyr::select(barcodes, UMAP1 = UMAP.1, UMAP2 = UMAP.2) |> 
```

    singler_combined_tumor = dplyr::if_else(
      singler_tumor_annotation_lumped %in% c("tumor_original", "tumor_new"), 
      "tumor", 
      singler_tumor_annotation_lumped
      ) |> 
      forcats::fct_relevel("tumor", after = 0) |> 
      forcats::fct_relevel("All remaining cell types", after = Inf),
    # make a label that is just tumor or normal
    tumor_normal_classification = dplyr::if_else(singler_combined_tumor == "tumor", "tumor", "normal") |> 
      forcats::fct_relevel("tumor", after = 0)
  )
```

```{r}
# read in marker genes table 
marker_genes_df <- readr::read_tsv(params$marker_genes_file, show_col_types = FALSE) |> 
  # account for genes being from multiple sources
  dplyr::select(cell_type, ensembl_gene_id, gene_symbol) |> 
  dplyr::distinct()

# get list of all cell types found
cell_types <- unique(marker_genes_df$cell_type)

# get the sum of expression of all genes for each cell type
gene_exp_df <- cell_types |> 
  purrr::map(\(type){
    calculate_sum_markers(marker_genes_df, type)
  }) |> 
  purrr::reduce(dplyr::inner_join, by = "barcodes")

# join sum expression columns with classification df 
classification_df <- classification_df |> 

>   dplyr::left_join(gene_exp_df)

Reviewer: Avoid warning
```suggestion
  dplyr::left_join(gene_exp_df, by = "barcodes")
```
